### PR TITLE
Better Edge Detection API

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
As discussed in #1297 It's not currently possible to implement your own custom edge detection kernels. 

This goes against the nature of the API and as such I deem it necessary to refactor the API to allow extensibility. This is a breaking change.

The API has been refactored to expose three structs representing three types of edge detection kernels.

- `EdgeDetectorKernel` Contains a single 2D XY kernel.
- `EdgeDetector2DKernel` Contains two separable 1D kernels. 
- `EdgeDetectorCompassKernel` Contains eight kernels representing each point on the compass. 

A new static class `KnownEdgeDetectorKernels` replaces the current enum `EdgeDetectionOperators`. 

Individual edge detection processors `SobelProcessor` etc have been replaced with versions representing each of the three kernel types.

The resulting code is much simpler and more powerful allowing users to implement any series of gradient operators they require.

### CODE REVIEW

File: src/ImageSharp/ImageSharp.csproj [hunk 1]
```diff
@@ -20,7 +20,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.SourceLink.GitHub"/>
+    <PackageReference Include="Microsoft.SourceLink.GitHub" />
```
Comment (dlemstra) [inline src/ImageSharp/ImageSharp.csproj:23]: Dp we not need `PrivateAssets="All"` here?

Comment (JimBobSquarePants) [reply src/ImageSharp/ImageSharp.csproj:23]: It's already marked private in the `Directory.Build.targets` file.

**Comment by JimBobSquarePants** [comment]
Perfect @dalingrin I’m glad I was able to improve matters.

**Comment by JimBobSquarePants** [comment]
@SixLabors/core I'm going to merge this unless anyone objects. I'm very happy with it.

**Pull request merged**